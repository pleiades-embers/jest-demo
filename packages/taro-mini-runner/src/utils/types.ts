import { IProjectBaseConfig, IMiniAppConfig } from '@tarojs/taro/types/compile'

export interface IFileType {
  // 样式文件后缀
  style: string
  // 脚本文件后缀
  script: string
  // 模板文件后缀
  templ: string
  // 配置文件后缀
  config: string
  // 【可选】渲染层脚本文件后缀，如微信小程序的 wxs，支付宝小程序的 sjs
  xs?: string
}

export type Func = (...args: any[]) => any

export interface IBuildConfig extends IProjectBaseConfig, IMiniAppConfig {
  isWatch: boolean
  mode: 'production' | 'development'
  port?: number
  buildAdapter: string
  nodeModulesPath: string
  quickappJSON: any
  isBuildPlugin: boolean
  isBuildQuickapp: boolean
  isSupportRecursive: boolean
  fileType: IFileType
  isSupportXS: boolean
  globalObject: string
  modifyWebpackChain: Func
  modifyBuildAssets: Func
  modifyMiniConfigs: Func
  modifyComponentConfig: Func
  onCompilerMake: Func
  onParseCreateElement: Func
  onWebpackChainReady: Func
  onBuildFinish: Func
  framework: string
  baseLevel: number
  runtimePath?: string | string[]
  taroComponentsPath?: string
  blended?: boolean
  isBuildNativeComp?: boolean
  prerender?: any
  template: any
}

export interface IOption {
  [key: string]: any
}
