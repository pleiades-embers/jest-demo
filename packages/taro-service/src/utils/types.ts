export interface IPaths {
    /**
     * 当前命令执行的目录，如果是 build 命令则为当前项目路径
     */
    appPath: string
    /**
     * 当前项目配置目录，如果 init 命令，则没有此路径
     */
    configPath: string
    /**
     * 当前项目源码路径
     */
    sourcePath: string
    /**
     * 当前项目输出代码路径
     */
    outputPath: string
    /**
     * 当前项目所用的 node_modules 路径
     */
    nodeModulesPath: string
  }
  