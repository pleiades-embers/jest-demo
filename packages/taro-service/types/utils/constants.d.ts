export declare const CONFIG_DIR_NAME = "config";
export declare const DEFAULT_CONFIG_FILE = "index";
export declare const PRESET_PREFIX = "@tarojs/preset-";
export declare const PLUGIN_PREFIX = "@tarojs/plugin-";
