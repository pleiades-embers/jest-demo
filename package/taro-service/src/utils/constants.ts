export const CONFIG_DIR_NAME = "config";
export const DEFAULT_CONFIG_FILE = "index";

export const PRESET_PREFIX = "@tarojs/preset-";
export const PLUGIN_PREFIX = "@tarojs/plugin-";
