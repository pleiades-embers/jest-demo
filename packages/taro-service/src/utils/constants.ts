export const CONFIG_DIR_NAME = "config";
export const DEFAULT_CONFIG_FILE = "index";

export const PRESET_PREFIX = "@tarojs/preset-";
export const PLUGIN_PREFIX = "@tarojs/plugin-";

export const IS_EVENT_HOOK = /^on/
export const IS_ADD_HOOK = /^add/
export const IS_MODIFY_HOOK = /^modify/


export enum PluginType {
    Preset='preset',
    Plugin='plugin',
}