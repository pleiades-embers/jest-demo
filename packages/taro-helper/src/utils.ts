import * as fs from 'fs-extra';
import * as path from 'path';
import * as findWorkspaceRoot from 'find-yarn-workspace-root';
import chalk from 'chalk';

import { processTypeEnum, processTypeMap } from './constants';

export const JS_EXT: string[] = ['.js', '.jsx'];
export const TS_EXT: string[] = ['.ts', '.tsx'];
export const SCRIPT_EXT: string[] = JS_EXT.concat(TS_EXT);

export function resolveMainFilePath(p: string, ext = SCRIPT_EXT): string {
  const realPath = p;
  const taroEnv = process.env.TARO_ENV;

  for (let i = 0; i < ext.length; i++) {
    const item = ext[i];
    if (taroEnv) {
      if (fs.existsSync(`${p}.${taroEnv}${item}`)) {
        return `${p}.${taroEnv}${item}`;
      }
      if (fs.existsSync(`${p}.${path.sep}index.${taroEnv}${item}`)) {
        return `${p}${path.sep}index.${taroEnv}${item}`;
      }
      if (
        fs.existsSync(`${p.replace(/\/index$/, `.${taroEnv}/index`)}${item}`)
      ) {
        return `${p.replace(/\/index$/, `.${taroEnv}/index`)}${item}`;
      }
    }
    if (fs.existsSync(`${p}${item}`)) {
      return `${p}${item}`;
    }
    if (fs.existsSync(`${p}${path.sep}index${item}`)) {
      return `${p}${path.sep}index${item}`;
    }
  }
  return realPath;
}

export function resolveScriptPath(p: string): string {
  return resolveMainFilePath(p);
}

// 默认导出兼容获取
export const getModuleDefaultExport = (exports) =>
  exports.__esModule ? exports.default : exports;

export function printLog(
  type: processTypeEnum,
  tag: string,
  filePath?: string
) {
  const typeShow = processTypeMap[type];
  const tagLen = tag.replace(/[\u0391-\uFFE5]/g, 'aa').length;
  const tagFormatLen = 8;
  if (tagLen < tagFormatLen) {
    const rightPadding = new Array(tagFormatLen - tagLen + 1).join(' ');
    tag += rightPadding;
  }
  const padding = '';
  filePath = filePath || '';
  if (typeof typeShow.color === 'string') {
    console.log(
      chalk[typeShow.color](typeShow.name),
      padding,
      tag,
      padding,
      filePath
    );
  }else {
    console.log(typeShow.color(typeShow.name),padding,tag,padding,filePath);
  }
}

export function recursiveFindNodeModules(
  filePath: string,
  lastFindPath?: string
): string {
  const dirname = path.dirname(filePath);
  const workspaceRoot = findWorkspaceRoot(dirname);
  const nodeModules = path.join(workspaceRoot || dirname, 'node_modules');
  if (fs.existsSync(nodeModules)) {
    return nodeModules;
  }
  if (dirname.split(path.sep).length <= 1) {
    printLog(processTypeEnum.ERROR,`在${dirname}目录下`,'为找到')
    return nodeModules;
  }
  return recursiveFindNodeModules(dirname, filePath);
}
