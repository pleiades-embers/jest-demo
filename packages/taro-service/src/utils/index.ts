import { PluginItem } from '@tarojs/taro/types/compile';
import * as path from 'path';
import { merge } from 'lodash';
import * as resolve from 'resolve';
import { PluginType } from './constants';
import { IPlugin } from './types';
import { getModuleDefaultExport } from '@tarojs/helper';

export const isNpmPkg: (name: string) => boolean = (name) =>
  !/^(\.|\/)/.test(name);

export function getPLuginPath(pluginPath: string) {
  if (isNpmPkg(pluginPath) || path.isAbsolute(pluginPath)) return pluginPath;
  throw new Error('plugin 和 presets 配置必须为绝对路径包名');
}

export function convertPluginsToObject(items: PluginItem[]) {
  return () => {
    const obj = {};
    if (Array.isArray(items)) {
      items.forEach((item) => {
        if (typeof item === 'string') {
          const name = getPLuginPath(item);
          obj[name] = null;
        } else if (Array.isArray(item)) {
          const name = getPLuginPath(item[0]);
          obj[name] = item[1];
        }
      });
    }
    return obj;
  };
}

export function mergePlugins(dist: PluginItem[], src: PluginItem[]) {
  return () => {
    const srcObj = convertPluginsToObject(src)();
    const distObj = convertPluginsToObject(dist)();
    return merge(srcObj, distObj);
  };
}

export function resolvePresetsOrPlugins(
  root: string,
  args,
  type: PluginType
): IPlugin[] {
  return Object.keys(args).map((item) => {
    const fPath = resolve.sync(item, {
      basedir: root,
      extensions: ['js', '.ts']
    });
    return {
      id: fPath,
      path: fPath,
      type,
      opts:args[item]||[],
      apply(){
          return getModuleDefaultExport(require(fPath))
      }
    };
  });
}
