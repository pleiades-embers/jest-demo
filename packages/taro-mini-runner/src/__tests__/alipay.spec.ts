import { compile } from './utils/compiler'

describe('Compiler', () => {
  test('should compile', async () => {
    const { stats } = await compile('react', {
      buildAdapter: 'alipay'
    })

    // const assets = stats.toJson().assets || []
    expect(stats).toMatchSnapshot()
  })
})
