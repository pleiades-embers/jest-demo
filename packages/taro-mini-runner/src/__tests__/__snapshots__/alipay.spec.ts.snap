// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compiler should compile 1`] = `1`;
